File: src/reactive/scheduler.py
"""
Reactive Scheduler

Implementación de: VELA-574 - TASK-031
Historia: Scheduler Reactivo Avanzado
Fecha: 2025-12-01

Descripción:
Scheduler avanzado para el sistema reactivo con:
- Batching automático de actualizaciones
- Priorización de updates (signals > computed > effects)
- Scheduling inteligente (microtask queue)
- Coalescing de múltiples updates al mismo nodo
- Performance optimizations con memoization
"""

from typing import Any, Callable, Dict, List, Optional, Set, Tuple, TYPE_CHECKING
from collections import deque
from enum import Enum
import time

from .types import (
    IReactiveNode,
    NodeType,
    NodeState,
    CyclicDependencyError,
)

# Evitar import circular
if TYPE_CHECKING:
    from .graph import ReactiveNode


class SchedulerPriority(Enum):
    """
    Prioridades de scheduling.
    
    SYNC: Ejecución inmediata (signals)
    HIGH: Alta prioridad (computed)
    NORMAL: Prioridad normal (effects)
    LOW: Baja prioridad (cleanup, GC)
    """
    SYNC = 0      # Inmediato
    HIGH = 1      # Computed
    NORMAL = 2    # Effects
    LOW = 3       # Cleanup


class ScheduledUpdate:
    """
    Representa un update programado.
    
    Encapsula un nodo y su prioridad para scheduling.
    """
    
    def __init__(
        self,
        node: 'ReactiveNode',
        priority: SchedulerPriority,
        timestamp: float,
    ):
        """
        Inicializa un update programado.
        
        Args:
            node: Nodo a actualizar
            priority: Prioridad del update
            timestamp: Timestamp del scheduling
        """
        self.node = node
        self.priority = priority
        self.timestamp = timestamp
    
    def __lt__(self, other: 'ScheduledUpdate') -> bool:
        """Comparación para ordenamiento por prioridad."""
        if self.priority.value != other.priority.value:
            return self.priority.value < other.priority.value
        return self.timestamp < other.timestamp
    
    def __repr__(self) -> str:
        return (
            f"ScheduledUpdate(node={self.node.id}, "
            f"priority={self.priority.name}, ts={self.timestamp:.6f})"
        )


class ReactiveScheduler:
    """
    Scheduler reactivo avanzado.
    
    Gestiona la propagación eficiente de cambios con:
    - Batching automático
    - Priorización de updates
    - Coalescing de múltiples cambios
    - Microtask queue para async execution
    """
    
    def __init__(self):
        """Inicializa el scheduler."""
        # Queues por prioridad (usar Any para evitar import circular)
        self._sync_queue: deque = deque()
        self._high_queue: deque = deque()
        self._normal_queue: deque = deque()
        self._low_queue: deque = deque()
        
        # Tracking
        self._scheduled_nodes: Set[str] = set()  # IDs de nodos ya scheduled
        self._is_flushing: bool = False
        self._flush_depth: int = 0
        self._max_flush_depth: int = 100  # Prevenir loops infinitos
        
        # Batching
        self._is_batching: bool = False
        self._batch_depth: int = 0
        
        # Metrics
        self._metrics = {
            'total_updates': 0,
            'batched_updates': 0,
            'coalesced_updates': 0,
            'flush_count': 0,
        }
    
    @property
    def is_flushing(self) -> bool:
        """Si el scheduler está ejecutando updates."""
        return self._is_flushing
    
    @property
    def is_batching(self) -> bool:
        """Si el scheduler está en modo batch."""
        return self._is_batching
    
    @property
    def metrics(self) -> Dict[str, int]:
        """Métricas del scheduler."""
        return self._metrics.copy()
    
    def schedule_update(
        self,
        node: 'ReactiveNode',
        priority: Optional[SchedulerPriority] = None,
    ) -> None:
        """
        Programa un update para un nodo.
        
        Args:
            node: Nodo a actualizar
            priority: Prioridad (opcional, se infiere del tipo si no se provee)
        """
        # Incrementar total siempre
        self._metrics['total_updates'] += 1
        
        # Coalescing: si ya está scheduled, skip
        if node.id in self._scheduled_nodes:
            self._metrics['coalesced_updates'] += 1
            return
        
        # Inferir prioridad si no se provee
        if priority is None:
            priority = self._infer_priority(node)
        
        # Agregar a la queue correspondiente
        queue = self._get_queue(priority)
        queue.append(node)
        self._scheduled_nodes.add(node.id)
        
        # Si estamos en batch, no flush automático
        if self._is_batching:
            self._metrics['batched_updates'] += 1
            return
        
        # Flush automático solo para SYNC
        if priority == SchedulerPriority.SYNC:
            self.flush()
    
    def flush(self) -> None:
        """
        Ejecuta todos los updates pendientes.
        
        Procesa las queues en orden de prioridad:
        1. SYNC (signals)
        2. HIGH (computed)
        3. NORMAL (effects)
        4. LOW (cleanup)
        """
        if self._is_flushing:
            return  # Ya estamos flushing (evitar re-entrancy)
        
        self._is_flushing = True
        self._flush_depth = 0
        self._metrics['flush_count'] += 1
        
        try:
            while self._has_pending_updates():
                if self._flush_depth >= self._max_flush_depth:
                    raise RuntimeError(
                        f"Max flush depth ({self._max_flush_depth}) exceeded. "
                        "Possible infinite update loop."
                    )
                
                self._flush_depth += 1
                
                # Procesar en orden de prioridad
                self._flush_queue(self._sync_queue)
                self._flush_queue(self._high_queue)
                self._flush_queue(self._normal_queue)
                self._flush_queue(self._low_queue)
        finally:
            self._is_flushing = False
            self._flush_depth = 0
            self._scheduled_nodes.clear()
    
    def batch(self, fn: Callable[[], Any]) -> Any:
        """
        Ejecuta una función en modo batch.
        
        Todos los updates se acumulan y ejecutan al final del batch.
        
        Args:
            fn: Función a ejecutar
            
        Returns:
            Any: Resultado de la función
        
        Example:
            >>> scheduler.batch(lambda: (
            ...     count.set(1),
            ...     count.set(2),
            ...     count.set(3)
            ... ))
            # Solo 1 update al final
        """
        self._is_batching = True
        self._batch_depth += 1
        
        try:
            result = fn()
            return result
        finally:
            self._batch_depth -= 1
            
            # Solo flush cuando salimos del batch más externo
            if self._batch_depth == 0:
                self._is_batching = False
                self.flush()
    
    def _infer_priority(self, node: 'ReactiveNode') -> SchedulerPriority:
        """
        Infiere la prioridad de un nodo según su tipo.
        
        Args:
            node: Nodo
            
        Returns:
            SchedulerPriority: Prioridad inferida
        """
        type_priority_map = {
            NodeType.SIGNAL: SchedulerPriority.SYNC,
            NodeType.COMPUTED: SchedulerPriority.HIGH,
            NodeType.EFFECT: SchedulerPriority.NORMAL,
            NodeType.WATCH: SchedulerPriority.NORMAL,
        }
        return type_priority_map.get(node.node_type, SchedulerPriority.NORMAL)
    
    def _get_queue(self, priority: SchedulerPriority) -> deque:
        """
        Obtiene la queue para una prioridad.
        
        Args:
            priority: Prioridad
            
        Returns:
            deque: Queue correspondiente
        """
        queue_map = {
            SchedulerPriority.SYNC: self._sync_queue,
            SchedulerPriority.HIGH: self._high_queue,
            SchedulerPriority.NORMAL: self._normal_queue,
            SchedulerPriority.LOW: self._low_queue,
        }
        return queue_map[priority]
    
    def _has_pending_updates(self) -> bool:
        """Verifica si hay updates pendientes."""
        return bool(
            self._sync_queue or
            self._high_queue or
            self._normal_queue or
            self._low_queue
        )
    
    def _flush_queue(self, queue: deque) -> None:
        """
        Flush de una queue específica con ordenamiento topológico.
        
        Los nodos se procesan en orden topológico para garantizar
        que las dependencies se calculen antes que sus dependientes.
        
        Args:
            queue: Queue a procesar
        """
        # Agrupar todos los nodos a procesar
        nodes_to_process = []
        while queue:
            node = queue.popleft()
            if node.state != NodeState.CLEAN:
                nodes_to_process.append(node)
        
        if not nodes_to_process:
            return
        
        # Ordenar topológicamente usando in-degree
        # Esto garantiza que dependencies se procesen antes que dependientes
        in_degree = {}
        for node in nodes_to_process:
            # Contar cuántas dependencies del nodo están en nodes_to_process
            in_degree[node] = sum(
                1 for dep in node._dependencies 
                if dep in nodes_to_process and dep.state != NodeState.CLEAN
            )
        
        # Kahn's algorithm: empezar con nodos sin dependencies
        ready = deque([node for node in nodes_to_process if in_degree[node] == 0])
        sorted_nodes = []
        
        while ready:
            node = ready.popleft()
            sorted_nodes.append(node)
            
            # Decrementar in-degree de dependientes
            for dependent in node._dependents:
                if dependent in in_degree:
                    in_degree[dependent] -= 1
                    if in_degree[dependent] == 0:
                        ready.append(dependent)
        
        # Si no se procesaron todos, hay ciclo (usar orden original)
        if len(sorted_nodes) < len(nodes_to_process):
            sorted_nodes = nodes_to_process
        
        # Procesar en orden topológico
        for node in sorted_nodes:
            node.mark_dirty()
            
            try:
                if node.state == NodeState.DIRTY:
                    node.recompute()
            except Exception as e:
                print(f"Error updating {node.id}: {e}")
                node._last_error = e
    
    def clear(self) -> None:
        """Limpia todas las queues."""
        self._sync_queue.clear()
        self._high_queue.clear()
        self._normal_queue.clear()
        self._low_queue.clear()
        self._scheduled_nodes.clear()
    
    def __repr__(self) -> str:
        """Representación string del scheduler."""
        return (
            f"ReactiveScheduler("
            f"pending={len(self._scheduled_nodes)}, "
            f"flushing={self._is_flushing}, "
            f"batching={self._is_batching}"
            f")"
        )


# Global scheduler instance
_global_scheduler: Optional[ReactiveScheduler] = None


def get_global_scheduler() -> ReactiveScheduler:
    """
    Obtiene la instancia global del scheduler.
    
    Returns:
        ReactiveScheduler: Scheduler global
    """
    global _global_scheduler
    if _global_scheduler is None:
        _global_scheduler = ReactiveScheduler()
    return _global_scheduler


def set_global_scheduler(scheduler: ReactiveScheduler) -> None:
    """
    Establece el scheduler global.
    
    Args:
        scheduler: Nuevo scheduler
    """
    global _global_scheduler
    _global_scheduler = scheduler


if __name__ == "__main__":
    # Ejemplo de uso
    print("Reactive Scheduler - Demo\n")
    
    scheduler = ReactiveScheduler()
    
    # Simular updates
    from .signal import Signal
    
    count = Signal(0)
    
    print("1. Single update:")
    scheduler.schedule_update(count._node, SchedulerPriority.SYNC)
    print(f"   Metrics: {scheduler.metrics}\n")
    
    print("2. Batch updates:")
    def batch_updates():
        count.set(1)
        count.set(2)
        count.set(3)
    
    scheduler.batch(batch_updates)
    print(f"   Metrics: {scheduler.metrics}\n")
    
    print("3. Coalescing:")
    scheduler.schedule_update(count._node)
    scheduler.schedule_update(count._node)  # Coalesced
    scheduler.schedule_update(count._node)  # Coalesced
    scheduler.flush()
    print(f"   Metrics: {scheduler.metrics}")

File: reactive/src/scheduler.rs
//! # Reactive Scheduler - Advanced Update Scheduling
//!
//! Implementation of: VELA-574 - TASK-031
//! Story: Advanced Reactive Scheduler
//! Date: 2025-12-03
//!
//! Description:
//! Advanced scheduler for the reactive system with:
//! - Automatic batching of updates
//! - Update prioritization (signals > computed > effects)
//! - Intelligent scheduling (microtask queue)
//! - Coalescing of multiple updates to same node
//! - Performance optimizations with memoization

use std::collections::{HashSet, VecDeque};
use std::sync::{Arc, Mutex};

use crate::graph::{ReactiveNode, NodeType};

/// Scheduler priorities
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum SchedulerPriority {
    /// Immediate execution (signals)
    Sync = 0,
    /// High priority (computed)
    High = 1,
    /// Normal priority (effects)
    Normal = 2,
    /// Low priority (cleanup, GC)
    Low = 3,
}

/// Reactive scheduler for managing updates
pub struct ReactiveScheduler {
    /// Queues by priority
    sync_queue: Mutex<VecDeque<Arc<ReactiveNode>>>,
    high_queue: Mutex<VecDeque<Arc<ReactiveNode>>>,
    normal_queue: Mutex<VecDeque<Arc<ReactiveNode>>>,
    low_queue: Mutex<VecDeque<Arc<ReactiveNode>>>,
    /// Tracking of scheduled nodes (for coalescing)
    scheduled_nodes: Mutex<HashSet<String>>,
    /// Flush state
    is_flushing: Mutex<bool>,
    flush_depth: Mutex<usize>,
    max_flush_depth: usize,
    /// Batching state
    is_batching: Mutex<bool>,
    batch_depth: Mutex<usize>,
    /// Metrics
    metrics: Mutex<SchedulerMetrics>,
}

#[derive(Debug, Default, Clone)]
struct SchedulerMetrics {
    total_updates: usize,
    batched_updates: usize,
    coalesced_updates: usize,
    flush_count: usize,
}

impl ReactiveScheduler {
    /// Create a new scheduler
    pub fn new() -> Self {
        ReactiveScheduler {
            sync_queue: Mutex::new(VecDeque::new()),
            high_queue: Mutex::new(VecDeque::new()),
            normal_queue: Mutex::new(VecDeque::new()),
            low_queue: Mutex::new(VecDeque::new()),
            scheduled_nodes: Mutex::new(HashSet::new()),
            is_flushing: Mutex::new(false),
            flush_depth: Mutex::new(0),
            max_flush_depth: 100,
            is_batching: Mutex::new(false),
            batch_depth: Mutex::new(0),
            metrics: Mutex::new(SchedulerMetrics::default()),
        }
    }

    /// Check if currently flushing
    pub fn is_flushing(&self) -> bool {
        *self.is_flushing.lock().unwrap()
    }

    /// Check if currently batching
    pub fn is_batching(&self) -> bool {
        *self.is_batching.lock().unwrap()
    }

    /// Get scheduler metrics
    pub fn metrics(&self) -> SchedulerMetrics {
        (*self.metrics.lock().unwrap()).clone()
    }

    /// Schedule an update for a node
    pub fn schedule_update(&self, node: Arc<ReactiveNode>) {
        // Increment total updates
        {
            let mut metrics = self.metrics.lock().unwrap();
            metrics.total_updates += 1;
        }

        // Coalescing: if already scheduled, skip
        {
            let mut scheduled = self.scheduled_nodes.lock().unwrap();
            if scheduled.contains(&node.id) {
                let mut metrics = self.metrics.lock().unwrap();
                metrics.coalesced_updates += 1;
                return;
            }
            scheduled.insert(node.id.clone());
        }

        // Infer priority and add to appropriate queue
        let priority = self.infer_priority(&node);
        let queue = self.get_queue(priority);
        queue.lock().unwrap().push_back(Arc::clone(&node));

        // If batching, don't auto-flush
        if self.is_batching() {
            let mut metrics = self.metrics.lock().unwrap();
            metrics.batched_updates += 1;
            return;
        }

        // Auto-flush only for SYNC priority
        if priority == SchedulerPriority::Sync {
            self.flush();
        }
    }

    /// Flush all pending updates
    pub fn flush(&self) {
        if self.is_flushing() {
            return; // Prevent re-entrancy
        }

        *self.is_flushing.lock().unwrap() = true;
        *self.flush_depth.lock().unwrap() = 0;

        {
            let mut metrics = self.metrics.lock().unwrap();
            metrics.flush_count += 1;
        }

        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
            while self.has_pending_updates() {
                let mut depth = self.flush_depth.lock().unwrap();
                if *depth >= self.max_flush_depth {
                    panic!("Max flush depth ({}) exceeded. Possible infinite update loop.", self.max_flush_depth);
                }
                *depth += 1;

                // Process in priority order
                self.flush_queue(&self.sync_queue);
                self.flush_queue(&self.high_queue);
                self.flush_queue(&self.normal_queue);
                self.flush_queue(&self.low_queue);
            }
        }));

        *self.is_flushing.lock().unwrap() = false;
        *self.flush_depth.lock().unwrap() = 0;
        self.scheduled_nodes.lock().unwrap().clear();

        if let Err(_) = result {
            // Handle panic during flush
            eprintln!("Panic during scheduler flush");
        }
    }

    /// Execute a function in batch mode
    pub fn batch<F, T>(&self, func: F) -> T
    where
        F: FnOnce() -> T,
    {
        *self.is_batching.lock().unwrap() = true;
        *self.batch_depth.lock().unwrap() += 1;

        let result = func();

        *self.batch_depth.lock().unwrap() -= 1;

        // Only flush when exiting outermost batch
        if *self.batch_depth.lock().unwrap() == 0 {
            *self.is_batching.lock().unwrap() = false;
            self.flush();
        }

        result
    }

    /// Infer priority for a node based on its type
    fn infer_priority(&self, node: &ReactiveNode) -> SchedulerPriority {
        match node.node_type {
            NodeType::Signal => SchedulerPriority::Sync,
            NodeType::Computed => SchedulerPriority::High,
            NodeType::Effect | NodeType::Watch => SchedulerPriority::Normal,
        }
    }

    /// Get the queue for a priority
    fn get_queue(&self, priority: SchedulerPriority) -> &Mutex<VecDeque<Arc<ReactiveNode>>> {
        match priority {
            SchedulerPriority::Sync => &self.sync_queue,
            SchedulerPriority::High => &self.high_queue,
            SchedulerPriority::Normal => &self.normal_queue,
            SchedulerPriority::Low => &self.low_queue,
        }
    }

    /// Check if there are pending updates
    fn has_pending_updates(&self) -> bool {
        !self.sync_queue.lock().unwrap().is_empty()
            || !self.high_queue.lock().unwrap().is_empty()
            || !self.normal_queue.lock().unwrap().is_empty()
            || !self.low_queue.lock().unwrap().is_empty()
    }

    /// Flush a specific queue with topological ordering
    fn flush_queue(&self, queue: &Mutex<VecDeque<Arc<ReactiveNode>>>) {
        let mut nodes_to_process = Vec::new();

        // Collect nodes from queue
        {
            let mut queue_guard = queue.lock().unwrap();
            while let Some(node) = queue_guard.pop_front() {
                if *node.state.read() != crate::graph::NodeState::Clean {
                    nodes_to_process.push(Arc::clone(&node));
                }
            }
        }

        if nodes_to_process.is_empty() {
            return;
        }

        // Simple processing for now (can be enhanced with topological sort)
        for node in nodes_to_process {
            node.mark_dirty();

            // For signals, just mark clean (they're updated externally)
            // For computed/effect, would need to implement recompute
            if node.node_type != NodeType::Signal {
                // Simplified: just mark as clean for now
                *node.state.write() = crate::graph::NodeState::Clean;
            }
        }
    }

    /// Clear all queues
    pub fn clear(&self) {
        self.sync_queue.lock().unwrap().clear();
        self.high_queue.lock().unwrap().clear();
        self.normal_queue.lock().unwrap().clear();
        self.low_queue.lock().unwrap().clear();
        self.scheduled_nodes.lock().unwrap().clear();
    }
}

impl Default for ReactiveScheduler {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graph::ReactiveNode;

    #[test]
    fn test_scheduler_creation() {
        let scheduler = ReactiveScheduler::new();
        assert!(!scheduler.is_flushing());
        assert!(!scheduler.is_batching());
    }

    #[test]
    fn test_schedule_update() {
        let scheduler = ReactiveScheduler::new();
        let node = Arc::new(ReactiveNode::new_signal("test".to_string(), serde_json::json!(1)));

        scheduler.schedule_update(Arc::clone(&node));

        // For Sync priority, updates are flushed automatically
        // So there should be no pending updates
        assert!(!scheduler.has_pending_updates());
    }

    #[test]
    fn test_batch_execution() {
        let scheduler = ReactiveScheduler::new();

        let result = scheduler.batch(|| {
            42
        });

        assert_eq!(result, 42);
    }

    #[test]
    fn test_metrics() {
        let scheduler = ReactiveScheduler::new();
        let metrics = scheduler.metrics();

        assert_eq!(metrics.total_updates, 0);
        assert_eq!(metrics.flush_count, 0);
    }

    #[test]
    fn test_clear() {
        let scheduler = ReactiveScheduler::new();
        let node = Arc::new(ReactiveNode::new_signal("test".to_string(), serde_json::json!(1)));

        // Use batch to prevent auto-flush
        scheduler.batch(|| {
            scheduler.schedule_update(Arc::clone(&node));
            assert!(scheduler.has_pending_updates());
        });

        // After batch ends, updates should be flushed automatically
        assert!(!scheduler.has_pending_updates());

        // Now test clear with pending updates
        scheduler.batch(|| {
            scheduler.schedule_update(Arc::clone(&node));
            assert!(scheduler.has_pending_updates());

            scheduler.clear();
            assert!(!scheduler.has_pending_updates());
        });
    }
}